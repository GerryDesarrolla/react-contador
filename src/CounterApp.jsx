import PropTypes from "prop-types";
import {useState} from "react";
import {Theme} from "./components/Theme/Theme";

export const CounterApp = ({value}) => {
	const [contador, setContador] = useState(value);

	const restar = () => contador > 0 ? setContador(contador-1) : '';
	const sumar = () => setContador(contador+1);
	const reset = () => setContador(value);

	return (
		<>
			<Theme/>
			<div className="c-contador o-container">
				<h1 className="c-contador__titulo">Contador</h1>
				
				<div className="c-contador__conteo">
					<button aria-label="restar" className="o-boton o-boton__redondo o-boton__redondo--menos" onClick={restar} disabled={contador === 0}>-</button>
					<p className="c-contador__texto">{contador}</p>
					<button aria-label="sumar" className="o-boton o-boton__redondo o-boton__redondo--mas" onClick={sumar}>+</button>
				</div>
				
				<button aria-label="limpiar" className="o-boton o-boton__rectangular" onClick={reset} disabled={contador === 0}><span>Reiniciar</span></button>
			</div>
		</>
	)
}

CounterApp.propTypes = {
	value: PropTypes.number.isRequired
}
