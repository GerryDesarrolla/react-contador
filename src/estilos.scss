$font-family: Helvetica, Arial, sans-serif;

:root{
	--azul: #21232A;
	--blanco: #ffffff;
}

:root.light{
	--azul: #ffffff;
	--blanco: #21232A;
}

$azul: var(--azul);
$blanco: var(--blanco);

*{
	box-sizing: border-box;
}

html, body{
	background: $azul;
	color: $blanco;
	font: 400 1.3rem/1.6rem $font-family;
	margin: 0;
}

.o-container{
	align-items: center;
	display: flex;
	flex-direction: column;
	justify-content: center;
	margin: 0 auto;
	max-width: 1240px;
	min-height: 100vh;
	padding: 0 20px;
	width: 100%;
}

.o-boton{
	background: transparent;
	border: 2px solid $blanco;
	position: relative;
	transition: all 0.5s ease-in;

	&:hover{
		background: $blanco;

		&:disabled{
			background: transparent;
			
			&:before, &:after{
				background: $blanco;
			}
		}
	}

	&:disabled{
		opacity: 0.1;
	}

	&__redondo{
		border-radius: 50%;
		font-size: 0;
		height: 50px;
		width: 50px;

		&:before, &:after{
			background: $blanco;
			content: '';
			display: block;
			left: 50%;
			position: absolute;
			top: 50%;
			transform: translate(-50%, -50%);
			transition: all 0.5s ease-in;
		}

		&:before{
			height: 2px;
			width: 20px;
		}

		&:hover{
			&:after, &:before{
				background: $azul;
			}
		}

		&--mas{
			&:after{
				height: 20px;
				width: 2px;
			}
		}
	}

	&__rectangular{
		border-radius: 4px;
		color: $blanco;
		font: 400 1.3rem/1.6rem $font-family;
		height: 50px;
		padding: 5px 30px;
		position: relative;

		&:before{
			background: $blanco;
			content: '';
			height: 100%;
			left: 0;
			position: absolute;
			top: 0;
			transform: none;
			transition: all 0.5s ease-in;
			width: 0;
		}

		&:hover{
			background: transparent;
			color: $azul;
			
			&:before{
				background: $blanco;
				width: 100%;
			}

			&:disabled{
				color: $blanco;
				
				&:before{
					display: none;
				}
			}
		}

		span{
			position: relative;
		}
	}
}

.c-contador{
	&__titulo{
		margin: 0 0 50px 0;
	}

	&__conteo{
		align-items: center;
		display: flex;
		gap: 20px;
		justify-content: center;
		margin: 0 0 30px 0;
	}

	&__texto{
		margin: 0;
		max-width: calc(100vw - 170px);
		overflow: hidden;
		text-overflow: ellipsis;
		white-space: nowrap;
	}
}
