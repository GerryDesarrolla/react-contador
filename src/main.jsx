import React from "react";
import ReactDOM from 'react-dom/client';
import "normalize/index.styl"
import "./estilos.scss";
import {CounterApp} from "./CounterApp";

ReactDOM.createRoot(document.getElementById('root')).render(
	<React.StrictMode>
		<CounterApp value={0}/>
	</React.StrictMode>
)
