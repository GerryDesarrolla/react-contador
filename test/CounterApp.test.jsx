import {fireEvent, render, screen} from "@testing-library/react";
import {CounterApp} from "../src/CounterApp";

describe('Pruebas sobre <CounterApp/>', () => {
	const numero = 0;
	
	test('Hacer match con el snapshot', () => {
		const {container} = render(
			<CounterApp value={numero}/>
		)
		
		expect(container).toMatchSnapshot();
	})
	
	test(`Muestra el valor inicial ${numero}`, () => {
		render(
			<CounterApp value={numero}/>
		);
		
		expect(screen.getByText(numero)).toBeTruthy();
	})
	
	test('Debe aumentar en 1 al dar clic en +', () => {
		render(
			<CounterApp value={numero}/>
		);
		
		fireEvent.click(screen.getByRole('button', {name: 'sumar'}));
		
		expect(screen.getByText(numero+1)).toBeTruthy()
	})
	
	test('Debe restar 1 al dar clic en -', () => {
		render(
			<CounterApp value={numero}/>
		);
		
		fireEvent.click(screen.getByRole('button', {name: 'restar'}));
		
		numero === 0
			? expect(screen.getByText(numero)).toBeTruthy()
			: expect(screen.getByText(2)).toBeTruthy()
	})
	
	test('Debe regresar a su estado inicial al presionar reset', () => {
		render(
			<CounterApp value={numero}/>
		);
		
		fireEvent.click(screen.getByRole('button', {name: 'sumar'}));
		fireEvent.click(screen.getByRole('button', {name: 'sumar'}));
		fireEvent.click(screen.getByRole('button', {name: 'sumar'}));
		fireEvent.click(screen.getByRole('button', {name: 'limpiar'}));
		
		expect(screen.getByText(numero)).toBeTruthy();
	})
})
