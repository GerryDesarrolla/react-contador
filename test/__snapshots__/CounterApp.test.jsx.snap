// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <CounterApp/> Hacer match con el snapshot 1`] = `
<div>
  <div
    class="c-contador o-container"
  >
    <label
      class="c-contador__tema"
      for="theme"
    >
      <input
        class="c-contador__checkbox"
        id="theme"
        type="checkbox"
      />
      <span />
    </label>
    <h1
      class="c-contador__titulo"
    >
      Contador
    </h1>
    <div
      class="c-contador__conteo"
    >
      <button
        class="o-boton o-boton__redondo o-boton__redondo--menos"
        disabled=""
      >
        -
      </button>
      <p
        class="c-contador__texto"
      >
        0
      </p>
      <button
        class="o-boton o-boton__redondo o-boton__redondo--mas"
      >
        +
      </button>
    </div>
    <button
      class="o-boton o-boton__rectangular"
      disabled=""
    >
      <span>
        Reiniciar
      </span>
    </button>
  </div>
</div>
`;
